### PR TITLE
feat: ci plugin support update when ci file change

### LINKED ISSUE #1161: :four_leaf_clover: `Proposal`: update plugin when ci files changed
## What Would You Like to Add? Why Is This Needed?
current CI related plugins don't update when CI files changes, since CI file is a important component in these plugins, we should update plugins when CI files's content change. 

## Design
- add CI files and remote CI files SHA  to state files
- compare SHA to check if we need to update these CI files

### PR DESCRIPTION
**Pull request by steinliber**
Signed-off-by: Meng JiaFeng <jiafeng.meng@merico.dev>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [ ] I have added relevant tests

## Description
- ci plugin will update file in scm when ci files change
- rename scm client method for better understandable
- move ci server related code to server module

## Related Issues
close #1161

### CODE REVIEW

File: pkg/util/scm/git/state.go [hunk 1]
```diff
@@ -22,10 +23,14 @@ func (f *RepoFileStatus) EncodeToGitHubContentOption(commitMsg string) *github.R
...
+func CalculateGitHubBlobSHA(fileContent []byte) string {
 	p := fmt.Sprintf("blob %d\x00", len(fileContent))
 	h := sha1.New()
 	h.Write([]byte(p))
 	h.Write([]byte(fileContent))
 	return hex.EncodeToString(h.Sum(nil))
 }
+
+func CalculateLocalFileSHA(fileContent []byte) string {
+	return base64.StdEncoding.EncodeToString(fileContent)
```
Comment (iyear) [inline pkg/util/scm/git/state.go:35]: For what reason is base64 used?

**Review by iyear — APPROVED** [review]
LGTM. I am not familar with gitlab library, so I only review the basic logic.

File: pkg/util/scm/git/state.go (hunk 1 above)
Comment (iyear) [reply pkg/util/scm/git/state.go:35]: If one of the files is an image, the state file will be very large.

File: internal/pkg/plugininstaller/ci/installer.go [hunk 2]
```diff
@@ -59,9 +60,8 @@ func DeleteCIFiles(options plugininstaller.RawOptions) error {
 	}
 	//3. delete ci files from git repo
 	commitInfo := &git.CommitInfo{
-		CommitMsg:    deleteCommitMsg,
-		CommitBranch: defaultBranch,
-		GitFileMap:   gitMap,
+		CommitMsg:  deleteCommitMsg,
+		GitFileMap: gitMap,
```
Comment (aFlyBird0) [inline internal/pkg/plugininstaller/ci/installer.go:64]: Missing `CommitBranch`?

File: internal/pkg/plugininstaller/ci/option.go [hunk 3]
```diff
@@ -18,7 +19,7 @@ import (
 )
 
 type CIConfig struct {
-	Type      ciRepoType             `validate:"oneof=jenkins github gitlab" mapstructure:"type"`
+	Type      server.CIServerType    `validate:"oneof=jenkins github gitlab" mapstructure:"type"`
```
Comment (aFlyBird0) [inline internal/pkg/plugininstaller/ci/option.go:22]: This name is more clear. 👍

File: internal/pkg/plugininstaller/ci/server/server.go [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
 	// Type return ci type
-	Type() ciRepoType
+	Type() CIServerType
 	// CIFilePath returns the file path of ci config file
 	// gitlab and jenkins is just a file, so we can just use filename
 	// but GitHub use directory, we should process this situation
 	// for GitHub: return ".github/workflows" or ".github/workflows/subFilename"
 	// for gitlab, jenkins: will ignore subFilename param
 	CIFilePath(subFilename ...string) string
 	// filterCIFilesFunc returns a filter function to select ci config file
```
Comment (aFlyBird0) [inline internal/pkg/plugininstaller/ci/server/server.go:16]: Please also change the function name in the comment. 
`// FilterCIFilesFunc`

File: internal/pkg/plugininstaller/ci/server/server.go [hunk 5]
```diff
@@ -1,25 +1,25 @@
...
-	filterCIFilesFunc() file.DirFIleFilterFunc
+	FilterCIFilesFunc() file.DirFIleFilterFunc
 	// getGitNameFunc returns a function to transform file path to git name of ci config file
```
Comment (aFlyBird0) [inline internal/pkg/plugininstaller/ci/server/server.go:18]: `// GetGitNameFunc`

**Review by aFlyBird0 — APPROVED** [review]
LGTM

**Pull request merged**